File: src/next_best_view/next_best_view/object_detection.py
#!/usr/bin/env python3
from threading import Lock
import sys

import cv2
import numpy as np
from ultralytics import YOLO
import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer
from message_filters import ApproximateTimeSynchronizer, Subscriber
from sensor_msgs.msg import Image

from kinova_action_interfaces.action import DetectObject


class ObjectDetectionNode(Node):
    def __init__(self):
        super().__init__("object_detection_node")

        self.debug: bool = True

        # TODO: replace with cv_bridge once it's fixed
        # Initialize CV Bridge
        # self.bridge: CvBridge = CvBridge()
        # Load YOLO model
        self.model: YOLO = YOLO("yolo11n.pt")

        # Create action server
        self._action_server = ActionServer(
            self,
            DetectObject,
            "detect_object",
            self.detect_object_callback,
        )

        # Latest synchronized images storage with thread safety
        self.latest_frames = None
        self.frames_lock = Lock()

        self._color_sub = Subscriber(self, Image, "/camera/color/image_raw")
        self._depth_sub = Subscriber(self, Image, "/camera/depth/image_raw")
        # self._color_sub = Subscriber(self, Image, "/wrist_mounted_camera/image")
        # self._depth_sub = Subscriber(self, Image, "/wrist_mounted_camera/depth_image")
        self._time_sync = ApproximateTimeSynchronizer(
            [
                self._color_sub,
                self._depth_sub,
            ],  # Synchronized subscribers
            30,
            0.5,  # defines the delay (in seconds) with which messages can be synchronized
        )
        self._time_sync.registerCallback(self.sync_callback)

        self.get_logger().info("Object Detection Node has been started")

    def sync_callback(self, color: Image, depth: Image):
        """Callback for synchronized RGB and depth images"""
        try:
            # Convert ROS Image messages to OpenCV format
            rgb_image = self._imgmsg_to_cv2(color, desired_encoding="rgb8")
            depth_image = self._imgmsg_to_cv2(depth, desired_encoding="16UC1")

            # Store synchronized frames with thread safety
            with self.frames_lock:
                self.latest_frames = {
                    "rgb": rgb_image,
                    "depth": depth_image,
                }

        except Exception as e:
            self.get_logger().error(f"Error processing synchronized frames: {str(e)}")

    async def detect_object_callback(self, goal_handle):
        """Action server callback to process detection requests"""
        self.get_logger().info("Received detection request")

        feedback_msg = DetectObject.Feedback()
        result = DetectObject.Result()

        # Get the target class from the goal
        target_class = goal_handle.request.target_class
        # TODO: add colors here

        # Get latest synchronized frames with thread safety
        with self.frames_lock:
            if self.latest_frames is None:
                raise RuntimeError("No synchronized frame data available")
            frames = self.latest_frames.copy()

        # try:

        # Run YOLOv8 detection on RGB image
        results = self.model(frames["rgb"])

        # Get image dimensions
        rgb_h, rgb_w = frames["rgb"].shape[:2]
        depth_h, depth_w = frames["depth"].shape[:2]

        # Calculate scaling factors
        scale_x = depth_w / rgb_w
        scale_y = depth_h / rgb_h

        # Get detections for requested object class
        detections = []

        for r in results:
            boxes = r.boxes
            for box in boxes:
                cls = int(box.cls[0])
                cls_name = self.model.names[cls]

                if target_class in cls_name:
                    conf = float(box.conf[0])
                    x1, y1, x2, y2 = box.xyxy[0].tolist()

                    # Calculate center point
                    rgb_center_x = int((x1 + x2) / 2)
                    rgb_center_y = int((y1 + y2) / 2)

                    center_x = int(rgb_center_x * scale_x)
                    center_y = int(rgb_center_y * scale_y)

                    # Ensure we're within depth image bounds
                    center_x = min(max(0, center_x), depth_w - 1)
                    center_y = min(max(0, center_y), depth_h - 1)

                    # Get depth at center point (convert to meters)
                    depth = (
                        float(np.average(frames["depth"][center_y, center_x])) / 1000.0
                    )
                    self.get_logger().info(f"{target_class} object depth {depth}m")

                    detections.append(
                        {
                            "confidence": conf,
                            "bbox": [x1, y1, x2, y2],
                            "center": [rgb_center_x, rgb_center_y],
                            "depth": depth,
                            "depth_center": [center_x, center_y],
                        }
                    )

        # Sort detections by confidence
        detections.sort(key=lambda x: x["confidence"], reverse=True)

        if detections:
            best_detection = detections[0]

            # Calculate 3D position
            # TODO: figure out if we have and adjust these
            fx = 360.01333
            fy = 360.01333
            cx = 243.87228
            cy = 137.9218444

            X = (best_detection["center"][0] - cx) * best_detection["depth"] / fx
            Y = (best_detection["center"][1] - cy) * best_detection["depth"] / fy
            Z = best_detection["depth"]

            # Set result
            result.success = True
            result.position.x = float(X)
            result.position.y = float(Y)
            result.position.z = float(Z)
            result.confidence = best_detection["confidence"]

            # Send feedback
            feedback_msg.processing_status = (
                f"Object detected at ({X:.2f}, {Y:.2f}, {Z:.2f}) "
                f"with confidence: {result.confidence:.2f}"
            )
            goal_handle.publish_feedback(feedback_msg)

        else:
            result.success = False
            result.confidence = 0.0
            feedback_msg.processing_status = f"No {target_class} found in image"
            goal_handle.publish_feedback(feedback_msg)

        goal_handle.succeed()

        if self.debug == True:  # Print debug images
            annotated_image = self._draw_detections(
                frames["rgb"].copy(), detections, target_class
            )

        # except Exception as e:
        #     self.get_logger().error(f"Detection failed: {str(e)}")
        #     result.success = False
        #     result.confidence = 0.0
        #     goal_handle.succeed()

        return result

    def _draw_detections(self, image, detections, target_class):
        """Draw bounding boxes and labels on the image"""
        image_with_boxes = image.copy()

        for det in detections:
            # Get bbox coordinates
            x1, y1, x2, y2 = map(int, det["bbox"])
            conf = det["confidence"]

            # Draw bounding box
            cv2.rectangle(image_with_boxes, (x1, y1), (x2, y2), (0, 255, 0), 2)

            # Draw label with confidence
            label = f"{target_class}: {conf:.2f}"
            cv2.putText(
                image_with_boxes,
                label,
                (x1, y1 - 10),
                cv2.FONT_HERSHEY_SIMPLEX,
                0.5,
                (0, 255, 0),
                2,
            )

            # Draw center point
            center_x, center_y = det["center"]
            cv2.circle(
                image_with_boxes, (int(center_x), int(center_y)), 4, (255, 0, 0), -1
            )

            image_with_boxes = cv2.cvtColor(image_with_boxes, cv2.COLOR_BGR2RGB)
            cv2.imwrite(f"{target_class}_detected.jpg", image_with_boxes)

        return image_with_boxes

    def _imgmsg_to_cv2(self, img_msg, desired_encoding):
        """
        This function is required to replace cv_bridge since it's not built with numpy 2.
        There is an error loading the module since it isn't compatible with the pyenv.
        Instead, we rewrite the function we use from that library to get rid of it until they fix.
        NOTE: I'm not sure if this actually does what it's supposed to do.

        "bgr8" from https://robotics.stackexchange.com/questions/95630/cv-bridge-throws-boost-import-error-in-python-3-and-ros-melodic
        "passthrough" from https://gist.github.com/Merwanski/39580a5fc276583b0921eb44dd91a61e
        """
        if desired_encoding == "rgb8":
            dtype = np.dtype("uint8")  # Hardcode to 8 bits...
            image_opencv = np.ndarray(
                shape=(
                    img_msg.height,
                    img_msg.width,
                    3,
                ),  # and three channels of data. Since OpenCV works with bgr natively, we don't need to reorder the channels.
                dtype=dtype,
                buffer=img_msg.data,
            )
            return image_opencv
        else:
            # Get the image dimensions
            height = img_msg.height
            width = img_msg.width

            # Get raw data from message
            raw_data = np.frombuffer(img_msg.data, dtype=np.uint8)

            # Since the data is 16-bit but stored in 8-bit bytes,
            # we need to reshape and reinterpret it
            raw_data = raw_data.reshape(-1, 2)  # Pair bytes together
            depth_data = raw_data[:, 0].astype(np.uint16) + (
                raw_data[:, 1].astype(np.uint16) << 8
            )

            # Reshape to 2D array with correct dimensions
            image_opencv = depth_data.reshape(height, width)

            return image_opencv


def main(args=None):
    rclpy.init(args=args)

    node = ObjectDetectionNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/next_best_view/src/object_detection.cpp
#include <memory>
#include <mutex>
#include <vector>
#include <string>

#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>
#include <message_filters/subscriber.h>
#include <message_filters/synchronizer.h>
#include <message_filters/sync_policies/approximate_time.h>
#include <sensor_msgs/msg/image.hpp>
#include <geometry_msgs/msg/transform_stamped.hpp>
#include <geometry_msgs/msg/point_stamped.hpp>
#include <tf2_ros/buffer.h>
#include <tf2_ros/transform_listener.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <opencv2/opencv.hpp>
#include <Eigen/Dense>

#include "kinova_action_interfaces/action/detect_object.hpp"
#include "next_best_view/yolo_detector.hpp"

using namespace std::placeholders;
using DetectObject = kinova_action_interfaces::action::DetectObject;
using GoalHandleDetectObject = rclcpp_action::ServerGoalHandle<DetectObject>;

const float METER_TO_MILLIMETER = 1000.0f;

class ObjectDetectionNode : public rclcpp::Node {
public:
    ObjectDetectionNode()
        : Node("object_detection_node"),
          tf_buffer_(this->get_clock()),  // Use node's clock
          tf_listener_(tf_buffer_) {      // Pass buffer directly
        debug_ = true;

        // Initialize YOLO model (placeholder)
        yolo_model_ = std::make_unique<YoloDetector>("yolo11n.pt");

        // Action server
        action_server_ = rclcpp_action::create_server<DetectObject>(
            this,
            "/next_best_view/detect_object",
            std::bind(&ObjectDetectionNode::handle_goal, this, _1, _2),
            std::bind(&ObjectDetectionNode::handle_cancel, this, _1),
            std::bind(&ObjectDetectionNode::handle_accepted, this, _1)
        );

        // Image subscribers with synchronization
        color_sub_.subscribe(this, "/camera/color/image_raw");
        depth_sub_.subscribe(this, "/camera/depth/image_raw");

        using SyncPolicy = message_filters::sync_policies::ApproximateTime<
            sensor_msgs::msg::Image, sensor_msgs::msg::Image>;
        synchronizer_ = std::make_shared<message_filters::Synchronizer<SyncPolicy>>(
            SyncPolicy(30), color_sub_, depth_sub_);  // Only queue size
        synchronizer_->registerCallback(
            std::bind(&ObjectDetectionNode::sync_callback, this, _1, _2));

        RCLCPP_INFO(this->get_logger(), "Object Detection Node has been started");
    }

private:
    bool debug_;
    std::unique_ptr<YoloDetector> yolo_model_;
    rclcpp_action::Server<DetectObject>::SharedPtr action_server_;
    tf2_ros::Buffer tf_buffer_;
    tf2_ros::TransformListener tf_listener_;
    message_filters::Subscriber<sensor_msgs::msg::Image> color_sub_;
    message_filters::Subscriber<sensor_msgs::msg::Image> depth_sub_;
    std::shared_ptr<message_filters::Synchronizer<
        message_filters::sync_policies::ApproximateTime<
            sensor_msgs::msg::Image, sensor_msgs::msg::Image>>> synchronizer_;

    struct Frames {
        cv::Mat bgr;
        cv::Mat depth;
    };
    std::optional<Frames> latest_frames_;
    std::mutex frames_mutex_;

    void sync_callback(const sensor_msgs::msg::Image::ConstSharedPtr& color,
                       const sensor_msgs::msg::Image::ConstSharedPtr& depth) {
        try {
            cv::Mat bgr_image = imgmsg_to_cv2(*color, "bgr8");
            cv::Mat depth_image = imgmsg_to_cv2(*depth, "16UC1");

            std::lock_guard<std::mutex> lock(frames_mutex_);
            latest_frames_ = Frames{bgr_image.clone(), depth_image.clone()};
        } catch (const std::exception& e) {
            RCLCPP_ERROR(this->get_logger(), "Error processing frames: %s", e.what());
        }
    }

    rclcpp_action::GoalResponse handle_goal(
        const rclcpp_action::GoalUUID& uuid,
        std::shared_ptr<const DetectObject::Goal> goal) {
        (void)uuid;
        RCLCPP_INFO(this->get_logger(), "Received detection request for %s",
                    goal->target_class.c_str());
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }

    rclcpp_action::CancelResponse handle_cancel(
        const std::shared_ptr<GoalHandleDetectObject> goal_handle) {
        (void)goal_handle;
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    void handle_accepted(const std::shared_ptr<GoalHandleDetectObject> goal_handle) {
        std::thread{[this, goal_handle]() {
            this->detect_object_callback(goal_handle);
        }}.detach();
    }

    void detect_object_callback(const std::shared_ptr<GoalHandleDetectObject> goal_handle) {
        const auto goal = goal_handle->get_goal();
        auto feedback = std::make_shared<DetectObject::Feedback>();
        auto result = std::make_shared<DetectObject::Result>();

        std::string target_class = goal->target_class;
        float target_view_point_distance = goal->target_view_point_distance;  // Direct use as float

        Frames frames;
        {
            std::lock_guard<std::mutex> lock(frames_mutex_);
            if (!latest_frames_) {
                RCLCPP_ERROR(this->get_logger(), "No synchronized frame data available");
                goal_handle->abort(result);
                return;
            }
            frames = *latest_frames_;
        }

        try {
            auto detections = yolo_object_detection(frames, target_class);
            std::sort(detections.begin(), detections.end(),
                      [](const auto& a, const auto& b) { return a.confidence > b.confidence; });

            if (!detections.empty()) {
                const auto& best_detection = detections[0];
                auto [X, Y, Z] = compute_3d_position(best_detection);

                try {
                    *result = camera_to_base_link_transform(X, Y, Z, target_view_point_distance);
                    result->confidence = best_detection.confidence;
                    result->success = true;

                    std::stringstream ss;
                    ss << "Object detected at base_link coordinates ("
                       << result->object_position.x << ", " << result->object_position.y << ", "
                       << result->object_position.z << ") ("
                       << result->view_position.position.x << ", " << result->view_position.position.y << ", "
                       << result->view_position.position.z << ") ("
                       << result->view_position.orientation.x << ", " << result->view_position.orientation.y << ", "
                       << result->view_position.orientation.z << ", " << result->view_position.orientation.w << ") "
                       << "with confidence: " << result->confidence;
                    feedback->processing_status = ss.str();
                    goal_handle->publish_feedback(feedback);
                } catch (const std::exception& e) {
                    RCLCPP_ERROR(this->get_logger(), "TF Transform failed: %s", e.what());
                    result->success = false;
                    result->confidence = 0.0;
                    feedback->processing_status = "Failed to transform object position to base_link";
                    goal_handle->publish_feedback(feedback);
                }
            } else {
                result->success = false;
                result->confidence = 0.0;
                feedback->processing_status = "No " + target_class + " found in image";
                goal_handle->publish_feedback(feedback);
            }

            if (debug_) {
                cv::Mat annotated = draw_detections(frames.bgr.clone(), detections, target_class);
                cv::imwrite(target_class + "_detected.jpg", annotated);
            }

            goal_handle->succeed(result);
        } catch (const std::exception& e) {
            RCLCPP_ERROR(this->get_logger(), "Detection failed: %s", e.what());
            result->success = false;
            result->confidence = 0.0;
            goal_handle->succeed(result);
        }
    }

    DetectObject::Result camera_to_base_link_transform(float X, float Y, float Z,
                                                       float target_view_point_distance) {
        DetectObject::Result result;

        geometry_msgs::msg::PointStamped object_in_camera_frame;
        object_in_camera_frame.point.x = X;
        object_in_camera_frame.point.y = Y;
        object_in_camera_frame.point.z = Z;

        geometry_msgs::msg::PointStamped object_view_point;
        object_view_point.point.x = X;
        object_view_point.point.y = Y;
        object_view_point.point.z = Z - target_view_point_distance;

        auto transform = tf_buffer_.lookupTransform("base_link", "camera_link",
                                                    rclcpp::Time(0));
        geometry_msgs::msg::PointStamped object_in_base_frame;
        geometry_msgs::msg::PointStamped object_view_point_in_base_frame;
        tf2::doTransform(object_in_camera_frame, object_in_base_frame, transform);
        tf2::doTransform(object_view_point, object_view_point_in_base_frame, transform);

        result.object_position = object_in_base_frame.point;
        result.view_position.position = object_view_point_in_base_frame.point;

        Eigen::Vector3f object_pos(
            object_in_base_frame.point.x,
            object_in_base_frame.point.y,
            object_in_base_frame.point.z
        );
        Eigen::Vector3f view_pos(
            object_view_point_in_base_frame.point.x,
            object_view_point_in_base_frame.point.y,
            object_view_point_in_base_frame.point.z
        );

        auto quaternion = compute_orientation(object_pos, view_pos);
        result.view_position.orientation.x = quaternion.x();
        result.view_position.orientation.y = quaternion.y();
        result.view_position.orientation.z = quaternion.z();
        result.view_position.orientation.w = quaternion.w();

        return result;
    }

    Eigen::Quaternionf compute_orientation(const Eigen::Vector3f& object_pos,
                                           const Eigen::Vector3f& view_pos) {
        Eigen::Vector3f Z = (object_pos - view_pos).normalized();
        Eigen::Vector3f Y = Eigen::Vector3f::UnitZ();

        Y = Y - Z * Z.dot(Y);
        float Y_norm = Y.norm();
        if (Y_norm < 1e-6f) {
            if (std::abs(Z.z()) > 0.99f) {
                Y = Eigen::Vector3f::UnitY();
            } else {
                Y = Eigen::Vector3f::UnitX();
            }
            Y = Y - Z * Z.dot(Y);
            Y.normalize();
        } else {
            Y.normalize();
        }

        Eigen::Vector3f X = Y.cross(Z).normalized();

        Eigen::Matrix3f rotation_matrix;
        rotation_matrix << X, Y, Z;

        return Eigen::Quaternionf(rotation_matrix);
    }

    struct Detection {
        float confidence;
        std::vector<float> bbox; // [x1, y1, x2, y2]
        std::vector<int> center; // [x, y]
        float depth;
        std::vector<int> depth_center; // [x, y]
    };

    std::vector<Detection> yolo_object_detection(const Frames& frames,
                                                const std::string& target_class) {
        std::vector<Detection> detections;
        auto results = yolo_model_->detect(frames.bgr);

        int bgr_h = frames.bgr.rows, bgr_w = frames.bgr.cols;
        int depth_h = frames.depth.rows, depth_w = frames.depth.cols;
        float scale_x = static_cast<float>(depth_w) / bgr_w;
        float scale_y = static_cast<float>(depth_h) / bgr_h;

        for (const auto& result : results) {
            if (result.class_name.find(target_class) != std::string::npos) {
                Detection det;
                det.confidence = result.confidence;
                det.bbox = result.bbox;
                det.center = {(int)(result.bbox[0] + result.bbox[2]) / 2,
                              (int)(result.bbox[1] + result.bbox[3]) / 2};
                det.depth_center = {(int)(det.center[0] * scale_x),
                                    (int)(det.center[1] * scale_y)};
                det.depth_center[0] = std::clamp(det.depth_center[0], 0, depth_w - 1);
                det.depth_center[1] = std::clamp(det.depth_center[1], 0, depth_h - 1);
                det.depth = frames.depth.at<uint16_t>(det.depth_center[1], det.depth_center[0]) / METER_TO_MILLIMETER;
                detections.push_back(det);
                RCLCPP_INFO(this->get_logger(), "%s detected at depth %f m",
                            target_class.c_str(), det.depth);
            }
        }
        return detections;
    }

    std::tuple<float, float, float> compute_3d_position(const Detection& object) {
        const float fx = 360.01333f, fy = 360.01333f;
        const float cx = 243.87228f, cy = 137.9218444f;

        float X = (object.depth_center[0] - cx) * object.depth / fx;
        float Y = (object.depth_center[1] - cy) * object.depth / fy;
        float Z = object.depth;

        return {X, Y, Z};
    }

    cv::Mat draw_detections(cv::Mat image, const std::vector<Detection>& detections,
                            const std::string& target_class) {
        for (const auto& det : detections) {
            cv::rectangle(image,
                          cv::Point(det.bbox[0], det.bbox[1]),
                          cv::Point(det.bbox[2], det.bbox[3]),
                          cv::Scalar(0, 255, 0), 2);
            std::string label = target_class + ": " + std::to_string(det.confidence);
            cv::putText(image, label, cv::Point(det.bbox[0], det.bbox[1] - 10),
                        cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 255, 0), 2);
            cv::circle(image, cv::Point(det.center[0], det.center[1]), 4,
                       cv::Scalar(255, 0, 0), -1);
        }
        cv::cvtColor(image, image, cv::COLOR_BGR2RGB);
        return image;
    }

    cv::Mat imgmsg_to_cv2(const sensor_msgs::msg::Image& img_msg, const std::string& encoding) {
        if (encoding == "bgr8") {
            return cv::Mat(img_msg.height, img_msg.width, CV_8UC3,
                           const_cast<unsigned char*>(img_msg.data.data())).clone();
        } else { // "16UC1"
            cv::Mat raw(img_msg.height, img_msg.width * 2, CV_8U,
                        const_cast<unsigned char*>(img_msg.data.data()));
            cv::Mat depth(img_msg.height, img_msg.width, CV_16U);
            for (size_t i = 0; i < img_msg.height; ++i) {
                for (size_t j = 0; j < img_msg.width; ++j) {
                    depth.at<uint16_t>(i, j) = (raw.at<uint8_t>(i, j * 2 + 1) << 8) +
                                               raw.at<uint8_t>(i, j * 2);
                }
            }
            return depth.clone();
        }
    }
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ObjectDetectionNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}